dali/test/op_graph.h

```diff
@@ -0,0 +1,147 @@
+namespace dali {
+namespace testing {
+
+struct Node;
+
+struct Edge {
+  Edge() = default;
+  Edge(Node *n, int index, bool explicit_GPU = false) : node(n), index(index), explicit_GPU(false) {}
+
+  Node *node = nullptr;
+  int index = 0;
+  bool explicit_GPU = false;
+
+  Edge gpu() const {
+    auto ret = *this;
+    ret.explicit_GPU = true;
+    return ret;
+  }
+};
+
+struct Node {
+  Node() = default;
+  Node(const Node &) = delete;
+  Node(Node &&) = default;
+
+  Node(std::string op_name, std::string node_name, int backend_mask)
+  : op_name(std::move(op_name)), node_name(std::move(node_name)), backend_mask(backend_mask) {}
+
+  template <typename... Inputs>
+  Node &operator()(Inputs&&... inputs) {
+    int junk[] = { (add_in(inputs), 0)... };
+    (void)junk;
+    return *this;
+  }
+
+  void add_in(Edge e) {
+    in.push_back(e);
+    if (e.index >= (int)e.node->out.size())
+      e.node->out.resize(e.index+1);
+    e.node->out[e.index] = Edge(this, in.size()-1);
+  }
+
+  Edge operator[](int i) {
+    if (i >= num_outputs) num_outputs = i+1;
+    return { this, i };
+  }
+
+  enum Backend {
+    CPU = 1,
+    GPU = 2,
+    AnyBackend = 3,
+  };
+
+  std::string op_name;
+  std::string node_name;
+  int backend_mask = 0;
+  int num_outputs = 0;
+  std::vector<Edge> in;
+  std::vector<Edge> out;
+};
+
+struct OpDAG {
+  Node &add(std::string op_name, std::string node_name, int backend_mask = Node::AnyBackend) {
+    Node node(std::move(op_name), std::move(node_name), backend_mask);
+    auto result = nodes.emplace(node.op_name, std::move(node));
+    assert(result.second && "Duplicate node");
+    return result.first->second;
+  }
+
+  Node &add(std::string op_name, int backend_mask = Node::AnyBackend) {
+    std::string node_name = op_name;
+    for (int suffix = 0; nodes.count(node_name); suffix++) {
+      node_name = op_name + std::to_string(suffix);
+    }
+    return add(std::move(op_name), std::move(node_name), backend_mask);
+  }
+
+  bool check_cycles(
+      const Node *n,
+      std::unordered_set<const Node *> &done,
+      std::unordered_set<const Node *> &in_progress) const {
+    if (done.count(n))
+      return true;
+    if (!in_progress.insert(n).second)
+      return false;
+    for (auto in : n->in)
+      if (!check_cycles(in.node, done, in_progress))
+        return false;
+    done.insert(n);
+    in_progress.erase(n);
+    return true;
+  }
+
+  bool validate(bool checkInputReferenced) const {
+    std::unordered_set<const Node *> done;
+    std::unordered_set<const Node *> in_progress;
+    if (!check_cycles(&out, done, in_progress))
+      return false;
+    if (checkInputReferenced)
+      return done.count(&in);
+    return true;
+  }
+
+  std::unordered_map<std::string, Node> nodes;
+  Node in{"", "__input", Node::AnyBackend};
+  Node out{"", "__output", Node::AnyBackend};
+};
+
+
+OpDAG SingleOpGraph(const std::string &op_name, int backend_mask = Node::AnyBackend, int num_inputs = 1, int num_outputs = 1) {
```

Comment:
Isn't `num_inputs` and `num_outputs` reundant? I mean, we can figure it out on `OpSpec` creating level and it seems, we don't need it earlier?

Reply:
Maybe we can make 0 a special value (the number of inputs/outputs will be inferred from arguments supplied to Run).